### PR TITLE
Enable calling the Glean SDK General API from Java

### PR DESCRIPTION
**Pull request by Dexterp37**
This changes the API in a few places:

- adds `@JvmOverloads` to public methods in Glean that  have default arguments;
- adds `@JvmOverloads` to the Configuration object's  public constructor.
- moves the `channel` parameter of the configuration object  earlier in the function signature, to make it more convenient  to call.

@pocmo does these changes make sense, or am I introducing footguns that might be obvious to you? I tested it by building locally and pointing FirefoxReality to that build, this seems to work. No compilation error, at least.

@mdboom for a sanity check :)

---

### Pull Request checklist

- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### After merge
- [ ] **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- [ ] **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### CODE REVIEW

File: components/service/glean/src/main/java/mozilla/components/service/glean/config/Configuration.kt [hunk 1]
```diff
@@ -14,45 +14,47 @@ import mozilla.components.service.glean.net.PingUploader
...
     // constructor and only initialized with a proper default when calling the primary
     // constructor from the secondary, public one, below.
+    @JvmOverloads
     constructor(
         serverEndpoint: String = DEFAULT_TELEMETRY_ENDPOINT,
+        channel: String? = null,
         maxEvents: Int = DEFAULT_MAX_EVENTS,
-        httpClient: PingUploader = ConceptFetchHttpUploader(lazy { HttpURLConnectionClient() }),
-        channel: String? = null
+        httpClient: PingUploader = ConceptFetchHttpUploader(lazy { HttpURLConnectionClient() })
```
Comment (badboy) [inline components/service/glean/src/main/java/mozilla/components/service/glean/config/Configuration.kt]: what's the reason for the change of order here?

File: components/service/glean/src/main/java/mozilla/components/service/glean/Glean.kt [hunk 2]
```diff
@@ -91,6 +91,7 @@ open class GleanInternalAPI internal constructor () {
      * as shared preferences
      * @param configuration A Glean [Configuration] object with global settings.
      */
+    @JvmOverloads
```
Comment (badboy) [inline components/service/glean/src/main/java/mozilla/components/service/glean/Glean.kt]: The JVM Overlords. 🙈 (sorry, had to make that joke, nothing to see here, move on!)

Comment (Dexterp37) [reply components/service/glean/src/main/java/mozilla/components/service/glean/Glean.kt]: Oh my. But that's a good one :D

File: components/service/glean/src/main/java/mozilla/components/service/glean/config/Configuration.kt (hunk 1 above)
Comment (Dexterp37) [reply components/service/glean/src/main/java/mozilla/components/service/glean/config/Configuration.kt]: The `JvmOverloads` is basically generating as many function as the number of defaults. So we'll have:

```java
Configuration(server)
Configuration(server, maxEvents)
Configuration(server, maxEvents, httpClient)
Configuration(server, maxEvents, httpClient, channel)
```

So, with the previous ordering, in order to set the channel, we'd need to provide *all* the arguments, including the `httpClient`. This means we'd need to Javify `ConceptFetchHttpUploader` as well. If we did that, to set the channel, we'd need to:

```java
Configuration("someserver", 200, new ConceptFetchHttpUpdate(new Kotlin.lazy(.....)), "release")
```

vs

```kotlin
Configuration().copy(channel="release)
```

in kotlin. We could probably consider making the `channel` the first parameter, since it's the one that would most likely change.

**Comment by pocmo** [comment]
> @pocmo does these changes make sense, or am I introducing footguns that might be obvious to you? I tested it by building locally and pointing FirefoxReality to that build, this seems to work. No compilation error, at least.

No, that is fine. For everything else in AC we do not guarantee/maintain any Java compatibility though.

**Comment by Dexterp37** [comment]
> 
> 
> > @pocmo does these changes make sense, or am I introducing footguns that might be obvious to you? I tested it by building locally and pointing FirefoxReality to that build, this seems to work. No compilation error, at least.
> 
> No, that is fine. For everything else in AC we do not guarantee/maintain any Java compatibility though.

Thanks for checking. That's a good point. In this specific case, given that we deprecated *service-telemetry*, I think we should. `FirefoxReality` is planning on using sync, and sync is using Glean for telemetry :)

**Comment by pocmo** [comment]
> That's a good point. In this specific case, given that we deprecated service-telemetry, I think we should. FirefoxReality is planning on using sync, and sync is using Glean for telemetry :)

If Glean would not have been developed we would have migrated `service-telemetry` to Kotlin too. :)

Firefox Reality already contains some Kotlin code - so I do not think its super critical. But I can understand that Glean may want broader support and it also is very minimal what's actually needed here.

File: components/service/glean/src/main/java/mozilla/components/service/glean/config/Configuration.kt (hunk 1 above)
Comment (mdboom) [reply components/service/glean/src/main/java/mozilla/components/service/glean/config/Configuration.kt]: Just to confirm: hopefully that doesn't break the API for existing (Kotlin) usages?

Comment (Dexterp37) [reply components/service/glean/src/main/java/mozilla/components/service/glean/config/Configuration.kt]: Not that I found, no. We use named arguments in [Fenix](https://github.com/mozilla-mobile/fenix/blob/8a9340fd9d3ab4e5d1f849cd1de8bd3a2bde4a78/app/src/main/java/org/mozilla/fenix/components/metrics/GleanMetricsService.kt#L402)/reference-browser. So we should be fine.

**Comment by Dexterp37** [comment]
> 
> 
> > That's a good point. In this specific case, given that we deprecated service-telemetry, I think we should. FirefoxReality is planning on using sync, and sync is using Glean for telemetry :)
> 
> If Glean would not have been developed we would have migrated `service-telemetry` to Kotlin too. :)

Eheh, I see :)

> Firefox Reality already contains some Kotlin code - so I do not think its super critical. But I can understand that Glean may want broader support and it also is very minimal what's actually needed here.

Yes, I expected many more changes. Looks like very trivial ones, at least for the General API.

**Comment by Dexterp37** [comment]
bors=pocmo,mdboom,badboy
bors r+

**Comment by Dexterp37** [comment]
bors retry
